src/interp.rs

```diff
@@ -35,12 +35,177 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 // SOFTWARE.
 
-use rustc::ty::Ty;
+use rustc::ty::{Ty, TyCtxt};
+use rustc::mir::{
+    START_BLOCK, Mir, BasicBlock, BasicBlockData, Statement, StatementKind,
+    Rvalue, Place, Terminator, TerminatorKind
+};
 
-// Stores a value represented as a vector of bytes, along with its type annotation.
-// It's important to keep track of a value's type while it is being evaluated,
-// so that we know how to interpret the bytes.
+use machine;
+use machine::{Machine, Address};
+use primitive::eval_operand;
+
+const MEMORY_CAPACITY: usize = 1024;
+
+/// Stores a value represented as a vector of bytes, along with its type
+/// annotation.  It's important to keep track of a value's type while it is
+/// being evaluated, so that we know how to interpret the bytes.
+#[derive(Debug, Clone)]
 pub struct TypedVal<'tcx> {
     pub ty: Ty<'tcx>,
     pub val: Vec<u8>,
 }
+
+/// Builds a VM and starts interpretation from the given MIR function, this is
+/// normally the main method.
+pub fn entry_point<'a, 'tcx>(mir: &'tcx Mir<'tcx>,
+                             tcx: TyCtxt<'a, 'tcx, 'tcx>) {
+    let m = machine::Memory::new(MEMORY_CAPACITY);
+    let mut vm = Machine::new(m, tcx);
+    eval_mir(mir, &mut vm);
+}
+
+/// Evaluates a function call, pushing a new stack frame.
+pub fn eval_mir<'a, 'tcx>(mir: &'tcx Mir<'tcx>,
+                          vm: &mut Machine<'a, 'tcx>) {
+    let frame = machine::Frame::new(mir);
+    vm.push(frame);
+    eval_basic_block(START_BLOCK, vm)
+}
+
+/// The easy bit. BasicBlocks are a list of instructions which are guaranteed to
+/// always be executed sequentially. We just need to iterate over the statement
+/// list and execute each one in order.
+fn eval_basic_block(block: BasicBlock,
+                    vm: &mut Machine) {
+    let BasicBlockData {
+        ref statements,
+        ref terminator,
+        is_cleanup: _
+    } = *vm.cur_frame().mir.basic_blocks().get(block).unwrap();
+
+    for statement in statements {
+        eval_statement(block, statement, vm)
+    }
+
+    if let Some(ref terminator) = *terminator {
+        eval_terminator(terminator, vm);
+    }
+}
+
+fn eval_statement<'a, 'tcx>(block: BasicBlock,
+                  statement: &Statement<'tcx>,
+                  vm: &mut Machine<'a, 'tcx>) {
+    let Statement {
```

Comment:
Could this be `let kind = statement.kind` or similar?

Reply:
That would be much clearer!

Reply:
Or even `match statement.kind` and kill the let.

Reply:
Fixed